File: 3.x.x/crates-utopia/detector/src/detector/framework.rs
use crate::detection::framework::{Astro, Framework as FrameworkDetection};
use crate::error::DetectorError;
use crate::input::Input;

/// PHP `Utopia\Detector\Detector\Framework`.
#[derive(Debug)]
pub struct Framework {
    inputs: Vec<Input>,
    options: Vec<Box<dyn FrameworkDetection>>,
    packager: String,
}

impl Framework {
    /// PHP `Framework::INPUT_FILE`.
    pub const INPUT_FILE: &'static str = "file";
    /// PHP `Framework::INPUT_PACKAGES`.
    pub const INPUT_PACKAGES: &'static str = "packages";

    /// PHP `__construct(string $packager = 'pnpm')`.
    #[must_use]
    pub fn new(packager: impl Into<String>) -> Self {
        Self {
            inputs: Vec::new(),
            options: Vec::new(),
            packager: packager.into(),
        }
    }

    /// PHP `addInput(string $content, string $type = '')`.
    pub fn add_input(
        &mut self,
        content: impl Into<String>,
        type_: impl Into<String>,
    ) -> Result<&mut Self, DetectorError> {
        let type_ = type_.into();
        if type_ != Self::INPUT_FILE && type_ != Self::INPUT_PACKAGES {
            return Err(DetectorError::InvalidInputType(type_));
        }
        self.inputs.push(Input::new(content, type_));
        Ok(self)
    }

    /// PHP `addOption(Detection $option)`.
    pub fn add_option(&mut self, option: impl FrameworkDetection + 'static) -> &mut Self {
        self.options.push(Box::new(option));
        self
    }

    /// PHP `detect()`.
    #[must_use]
    pub fn detect(&self) -> Option<Box<dyn FrameworkDetection>> {
        let files: Vec<&str> = self
            .inputs
            .iter()
            .filter(|input| input.type_ == Self::INPUT_FILE)
            .map(|input| input.content.as_str())
            .collect();
        let packages: Vec<&str> = self
            .inputs
            .iter()
            .filter(|input| input.type_ == Self::INPUT_PACKAGES)
            .map(|input| input.content.as_str())
            .collect();

        let mut matches: Vec<(usize, usize, usize)> = Vec::new();
        for (index, detector) in self.options.iter().enumerate() {
            let mut count = 0usize;
            for package_json in &packages {
                for package_needed in detector.get_packages() {
                    let needle = format!("\"{package_needed}\"");
                    if package_json.contains(&needle) {
                        count += 1;
                    }
                }
            }
            count += detector
                .get_files()
                .iter()
                .filter(|file| files.contains(&file.as_str()))
                .count();
            matches.push((index, count, detector.parent_count()));
        }

        let positive: Vec<(usize, usize, usize)> = matches
            .into_iter()
            .filter(|(_, count, _)| *count > 0)
            .collect();
        if positive.is_empty() {
            return None;
        }

        let highest = positive
            .iter()
            .map(|(_, count, _)| *count)
            .max()
            .unwrap_or(0);
        let mut winners: Vec<(usize, usize, usize)> = positive
            .into_iter()
            .filter(|(_, count, _)| *count == highest)
            .collect();

        let best_index = if winners.len() == 1 {
            winners[0].0
        } else {
            winners.sort_by(|a, b| a.2.cmp(&b.2));
            let astro = Astro::new().get_name();
            let first = winners[0].0;
            if self.options[first].get_name() == astro && winners.len() > 1 {
                winners[1].0
            } else {
                first
            }
        };

        let mut detected = self.options[best_index].box_clone();
        detected.set_packager(self.packager.clone());
        Some(detected)
    }
}

File: src/Detector/Framework.php
<?php

namespace Utopia\Detector\Detector;

use Utopia\Detector\Detection\Framework as FrameworkDetection;
use Utopia\Detector\Detection\Framework\Astro;
use Utopia\Detector\Detector;
use Utopia\Detector\Yaml;

class Framework extends Detector
{
    public const INPUT_FILE = 'file';
    public const INPUT_PACKAGES = 'packages';

    /**
     * @var array<FrameworkDetection>
     */
    protected array $options = [];

    protected string $packager = 'pnpm';

    public function __construct(string $packager = 'pnpm')
    {
        parent::__construct();
        $this->packager = $packager;
    }

    public function addInput(string $content, string $type = ''): self
    {
        if ($type !== self::INPUT_FILE && $type !== self::INPUT_PACKAGES) {
            throw new \InvalidArgumentException("Invalid input type '{$type}'");
        }

        parent::addInput($content, $type);

        return $this;
    }

    /**
     * Performs framework detection based on input patterns.
     *
     * @return FrameworkDetection|null The detected framework or null if no match is found.
     */
    public function detect(): ?FrameworkDetection
    {
        $files = array_filter($this->inputs, fn ($input) => $input['type'] === self::INPUT_FILE);
        $files = array_map(fn ($input) => $input['content'], $files);

        $packages = array_filter($this->inputs, fn ($input) => $input['type'] === self::INPUT_PACKAGES);
        $packages = array_map(fn ($input) => $input['content'], $packages);

        // List of frameworks with count of matches
        $frameworkMatches = [];

        // List of frameworks with count of parents
        // Helps decide framework priority when multiple frameworks got same matches count
        // (framework with least parents wins)
        $fameworkParents = [];

        foreach ($this->options as $detector) {
            $frameworkMatches[$detector->getName()] = 0;
            $fameworkParents[$detector->getName()] = 0;
        }

        foreach ($this->options as $detector) {
            // Check package-based detection
            foreach ($packages as $manifest) {
                foreach ($detector->getPackages() as $packageNeeded) {
                    if ($this->hasPackage($manifest, $packageNeeded)) {
                        $frameworkMatches[$detector->getName()] += 1;
                    }
                }
            }

            // Check path-based detection
            $matches = array_intersect($detector->getFiles(), $files);
            $frameworkMatches[$detector->getName()] += \count($matches);

            // Figure out how many classes detector extends using native PHP
            $parent = \get_parent_class($detector);
            while ($parent !== false) {
                $fameworkParents[$detector->getName()] += 1;
                $parent = \get_parent_class($parent);
            }
        }

        // Filter out frameworks with 0 matches
        $frameworkMatches = array_filter($frameworkMatches, fn ($count) => $count > 0);

        if (\count($frameworkMatches) <= 0) {
            return null;
        }

        // Sort for framework with most matches to be first
        arsort($frameworkMatches);

        // Filter out non-max matches
        $highestMatch = $frameworkMatches[\array_key_first($frameworkMatches)];
        $frameworkMatches = array_filter($frameworkMatches, fn ($count) => $count == $highestMatch);

        if (\count($frameworkMatches) === 1) {
            $bestFramework = \array_key_first($frameworkMatches);
        } else {
            $bestFrameworks = \array_keys($frameworkMatches);
            usort($bestFrameworks, fn ($a, $b) => $fameworkParents[$a] <=> $fameworkParents[$b]);

            $bestFramework = $bestFrameworks[0];

            // Trick to prevent Astro (library-agnostic framework) from detecting everywhere
            $astro = (new Astro())->getName();
            if (
                $bestFramework === $astro &&
                // and 2nd detection equally good as Astro
                $frameworkMatches[$astro] == $frameworkMatches[$bestFrameworks[1]]
            ) {
                $bestFramework = $bestFrameworks[1];
            }
        }

        foreach ($this->options as $detector) {
            if ($detector->getName() === $bestFramework) {
                $detector->setPackager($this->packager);
                return $detector;
            }
        }

        return null;
    }

    /**
     * Matches a dependency key in JSON manifests, where keys are always double
     * quoted, and in YAML ones, where a key may be unquoted, quoted, or inside
     * a flow mapping such as `dependencies: {jaspr: ^0.23.0}`.
     */
    protected function hasPackage(string $manifest, string $package): bool
    {
        if (\str_contains($manifest, '"'.$package.'"')) {
            return true;
        }

        return Yaml::hasKey($manifest, $package);
    }
}
